environment.yml

```diff
@@ -0,0 +1,56 @@
+name: ray
```

Comment:
Can we get rid of this file? What new dependencies does this PR introduce?

Reply:
Also, we should document the changes to the installation instructions under `doc/source/`

Reply:
We'll definitely remove this file before merging in. @gdoubleyew and I wanted to verify we had identical environments and are using conda tools that automatically discover any ```environment.yml``` file. We add four dependencies, none of which we want to add to the project permanently.

- ```flask```: Python web app framework. See comments above. Don't think we want to introduce it unless there is some other need.
- requests: Python library for making web requests. See comments above re REST. Purely for convenience and not sure we want to introduce.
- socat: forwards requests. We want to remove this dependency by using the Python gateway process itself to forward requests. 
- ipython: Because we were tired of installing ipython in our environments.

Reply:
It's a separate question, though, whether we might want to include something like an ```environment.yml``` file to make it easier for folks to install dependencies if they are using ```conda```. I resisted ```conda``` for a long time, but have found that it does make managing environments (especially those with compiled components) a lot easier. Er, assuming someone already went through the headaches of defining the environment.

Reply:
Conda environments came up in https://github.com/ray-project/ray/issues/585. I haven't really used conda environments, but it sounds like a lot of people like them.

As for the dependencies, it's fine to have some temporary (or even potentially non-temporary) dependencies. We can make it low-impact for users who are not using the gateway by only installing them with `pip install ray[external]` or something like that.

Reply:
Ah, that's helpful context. Creating a ```conda``` package isn't hard, but I don't think it's necessary now since it seems like the ```pip``` install is pretty robust (not the case with all projects with compiled code...) and ```pip``` plays nicely with ```conda``` for the most part.

In general, though, I've found ```conda``` environments to be much more stable than python virtual environments because they isolate more. A number of projects especially in the data science / ML space are starting to recommend ```conda``` (e.g., ```pytorch```).